src/main/java/io/cdap/plugin/gcp/dataplex/source/DataplexBatchSource.java

```diff
@@ -0,0 +1,514 @@
+      setExecutionSpec(executionSpec).setSpark(spark).build();
+
+    try (DataplexServiceClient dataplexServiceClient =
+      DataplexUtil.getDataplexServiceClient(config.getCredentials())) {
+      task = dataplexServiceClient.createTaskAsync(LakeName.newBuilder().setLake(config.getLake()).
+          setProject(config.tryGetProject()).setLocation(config.getLocation()).build(), task, task.getDescription())
+        .get();
+    }
+    return task.getDescription();
+  }
+
+  /**
+   * Format query for Cloud storage entities.
+   *
+   * @param entity           Dataplex entity
+   * @param isPreviewEnabled
+   * @return
+   */
+  private String formatQuery(Entity entity, boolean isPreviewEnabled) {
+    String queryTemplate = "select * from %s.%s %s";
+    StringBuilder condition = new StringBuilder();
+    if (!Strings.isNullOrEmpty(config.getFilter())) {
+      condition.append("where ").append(config.getFilter());
+    }
+
+    // In case of preview limit the job to 1000 records
+    if (isPreviewEnabled) {
+      condition.append(" LIMIT 1000;");
+    }
+    return String.format(queryTemplate, config.getZone(), entity.getId(), condition);
+  }
+
+  @Override
+  public void transform(KeyValue<Object, Object> input, Emitter<StructuredRecord> emitter)
+    throws IOException {
+    if (entity.getSystem().equals(StorageSystem.BIGQUERY)) {
+      StructuredRecord transformed = outputSchema == null ?
+        transformer.transform((GenericData.Record) input.getValue()) :
+        transformer.transform((GenericData.Record) input.getValue(), outputSchema);
+      emitter.emit(transformed);
+    } else {
+      emitter.emit((StructuredRecord) input.getValue());
+    }
+  }
+
+  /**
+   * Gets bucket from supplied configuration.
+   * <p>
+   * If the supplied configuration doesn't specify a bucket, a bucket will get auto created and configuration modified
+   * to auto-delete this bucket on completion.
+   *
+   * @param configuration Hadoop configuration instance.
+   * @param project       GCP projectId
+   * @param bigQuery      bigquery client
+   * @param credentials   GCP credentials
+   * @param bucketName    bucket Name
+   * @return Bucket name.
+   */
+  private String createBucket(Configuration configuration,
+                              String project,
+                              BigQuery bigQuery,
+                              Credentials credentials,
+                              String bucketName) {
+    String bucket = String.format(BQ_TEMP_BUCKET_NAME_TEMPLATE, bucketName);
+    // By default, this option is false, meaning the job can not delete the bucket.
+    configuration.setBoolean("fs.gs.bucket.delete.enable", true);
+
+    // the dataset existence is validated before, so this cannot be null
+    Dataset bigQueryDataset = bigQuery.getDataset(DatasetId.of(datasetProject, dataset));
+    GCPUtils.createBucket(GCPUtils.getStorage(project, credentials),
+      bucket, bigQueryDataset.getLocation(), null);
```

Comment:
Do we need to add CMEK as a config/runtime args `Encryption Key Name` as we do in BigQuery Source instead of passing it as null?

Reply:
also can we use `createBucket()` method here as well instead of directly using `GCPUtils#createBucket`? Since it has concurrency protection.

Reply:
CMEK will be added later, will use create bucket.